common/thorhelper/roxiehelper.ipp

```diff
@@ -59,8 +59,25 @@ enum TracingCategory
 class LogItem;
 interface IRoxieContextLogger : extends IContextLogger
 {
-    // Override base interface with versions that add prefix
-    // We could consider moving some or all of these down into IContextLogger
+    virtual StringBuffer &getLogPrefix(StringBuffer &ret) const = 0;
+    virtual bool isIntercepted() const = 0;
+    virtual void CTXLOGa(TracingCategory category, const LogMsgCategory & cat, const LogMsgJobInfo & job, LogMsgCode code, const char *prefix, const char *text) const = 0;
+    void CTXLOGae(IException *E, const char *file, unsigned line, const char *prefix, const char *format, ...) const __attribute__((format(printf, 6, 7)));
+    virtual void CTXLOGaeva(IException *E, const char *file, unsigned line, const char *prefix, const char *format, va_list args) const = 0;
+    virtual void CTXLOGl(LogItem *) const = 0;
+    virtual bool isBlind() const = 0;
+    virtual void gatherStats(CRuntimeStatisticCollection & merged) const = 0;
+    virtual StringBuffer &getStats(StringBuffer &s) const = 0;
+    virtual bool collectingDetailedStatistics() const = 0;
+};
+
+template <class INTERFACE>
+class CRoxieContextLogger : public CDefaultContextLogger<INTERFACE>
+{
+    public:
+    virtual StringBuffer &getLogPrefix(StringBuffer &ret) const = 0;
+    virtual void CTXLOGa(TracingCategory category, const LogMsgCategory & cat, const LogMsgJobInfo & job, LogMsgCode code, const char *prefix, const char *text) const = 0;
+    virtual void CTXLOGaeva(IException *E, const char *file, unsigned line, const char *prefix, const char *format, va_list args) const = 0;
```

Comment:
it doesn't look quite right that the above 3 pure virtuals are redeclared (as they will be inherited from the inherited implemented interface.

And I think CRoxieContextLogger does not need to be a template, it will be clearer if it is a concrete partial implementation of IRoxieContextLogger.
i.e.
```
class CRoxieContextLogger : public CDefaultContextLogger<IRoxieContextLogger>
```

Then this CRoxieContextLogger should be inherited where required ( i.e. rather than CRoxieContextLogger<IRoxieContextLogger>)

and similarly where IRoxieAgentContext is needed, there should be a concrete implementation:
```
class CRoxieAgentContextLogger : public CDefaultContextLogger<IRoxieAgentContext>
```
which CRoxieContextBase, CRoxieAgentContextLogger and IndirectAgentContext derived from.

Both concrete implementations want the same implementation of CTXLOGva and logOperatorExceptionVA, but it is cleaner if they either clone the few lines of code involved, or if they call out to helper code to call common code, e.g.:
```
    virtual void CTXLOGva(const LogMsgCategory & cat, const LogMsgJobInfo & job, LogMsgCode code, const char *format, va_list args) const override __attribute__((format(printf,5,0)))
    {
        CRoxieContextLoggerHelper::CTXLOGva(*this, cat, job, code, format, args);
    }
```